documentation/experimental_features.md

```diff
@@ -106,3 +115,174 @@ MODULEPATH:
 
 Use `bolt plan show <plan-name>` to view details and parameters for a specific plan.
 ```
+
+## `ResourceInstance` data type
+
+This feature was introduced in [Bolt
+2.10.0](https://github.com/puppetlabs/bolt/blob/master/CHANGELOG.md#bolt-2100-2020-05-18).
+
+Bolt has had a limited ability to interact with Puppet's Resource Abstraction Layer. Users
+could use the `apply` function to generate catalogs and return events, and the `get_resources`
+plan function can be used to query resources on target. The `ResourceInstance` data type is
+the first step in enabling plan authors to build resource-based logic into their plans to
+enable a discover-inspect-execute workflow for interacting with resources on remote systems.
+
+The `ResourceInstance` data type is used to store information about a single resource on a
+target, including its observed state, desired state, and any related events.
+
+> **Note::** The `ResourceInstance` data type does not interact with or modify resources in
+  any way. It is only used to store information about a resource.
+
+### Creating `ResourceInstance` objects
+
+#### `Target.set_resources()`
+
+The recommended way to create `ResourceInstance` objects is by setting them directly on
+a `Target` object using the `Target.set_resources` function. This function can be used to 
+set one or more resources on a target at a time. You can read more about this function in
+the [Bolt plan functions reference documentation](plan_functions.md#set_resources).
+
+The `Target.set_resources` function can set existing `ResourceInstance` objects on a target,
+or take hashes of parameters to create new `ResourceInstance` objects and automatically set
+them on a target.
+
+When setting resources using a hash of parameters, you can pass either a single hash or an
+array of hashes:
+
+> **Note:** When passing a data hash to `Target.set_resources`, the `target` parameter
+  is **optional**. If the `target` parameter is not specified, the function automatically
+  sets the target to the target the function is called on.
+
+> **Note:** If the `target` parameter is any target other than the one you are setting the
+  resource on, Bolt will raise an error.
+
+```ruby
+$init_hash = {
+  'target'        => Optional[Target],
+  'type'          => Variant[String[1], Type[Resource]],
+  'title'         => String[1],
+  'state'         => Optional[Hash[String[1], Data]],
+  'desired_state' => Optional[Hash[String[1], Data]],
+  'events'        => Optional[Array[Hash[String[1], Data]]]
+}
+
+$target.set_resources(
+  Variant[Hash[String[1], Any], Array[Hash[String[1], Any]]] init_hash
+)
+```
+
+When setting resources using existing `ResourceInstance`s, you can pass either a single
```

Comment:
```suggestion
When setting resources using existing `ResourceInstance`s, you can pass either a single
```
This is more of a style thing, but I think the plural `ResourceInstance`s is going to look weird once it's rendered on the site. Maybe rephrase to "existing `ResourceInstance` objects"